File: src/coreclr/tools/aot/ObjWriter/debugInfo/dwarf/dwarfAbbrev.cpp
//===---- dwarfAbbrev.cpp ---------------------------------------*- C++ -*-===//
//
// dwarf abbreviations implementation
//
// Licensed to the .NET Foundation under one or more agreements.
// The .NET Foundation licenses this file to you under the MIT license.
//
//===----------------------------------------------------------------------===//

#include "dwarfAbbrev.h"
#include "llvm/MC/MCContext.h"
#include "llvm/MC/MCObjectFileInfo.h"

namespace DwarfAbbrev {

void Dump(MCObjectStreamer *Streamer, uint16_t DwarfVersion, unsigned TargetPointerSize) {
  uint16_t DW_FORM_size;
  switch (TargetPointerSize) {
    case 1:
      DW_FORM_size = dwarf::DW_FORM_data1;
      break;
    case 2:
      DW_FORM_size = dwarf::DW_FORM_data2;
      break;
    case 4:
      DW_FORM_size = dwarf::DW_FORM_data4;
      break;
    case 8:
      DW_FORM_size = dwarf::DW_FORM_data8;
      break;
    default:
      assert(false && "Unexpected TargerPointerSize");
      return;
  }

  const uint16_t AbbrevTable[] = {
    CompileUnit,
        dwarf::DW_TAG_compile_unit, dwarf::DW_CHILDREN_yes,
        dwarf::DW_AT_producer, dwarf::DW_FORM_string,
        dwarf::DW_AT_language, dwarf::DW_FORM_data2,
        dwarf::DW_AT_stmt_list, (DwarfVersion >= 4 ? dwarf::DW_FORM_sec_offset : dwarf::DW_FORM_data4),
        0, 0,

    BaseType,
        dwarf::DW_TAG_base_type, dwarf::DW_CHILDREN_no,
        dwarf::DW_AT_name, dwarf::DW_FORM_strp,
        dwarf::DW_AT_encoding, dwarf::DW_FORM_data1,
        dwarf::DW_AT_byte_size, dwarf::DW_FORM_data1,
        0, 0,

    EnumerationType,
        dwarf::DW_TAG_enumeration_type, dwarf::DW_CHILDREN_yes,
        dwarf::DW_AT_name, dwarf::DW_FORM_strp,
        dwarf::DW_AT_type, dwarf::DW_FORM_ref4,
        dwarf::DW_AT_byte_size, dwarf::DW_FORM_data1,
        0, 0,

    Enumerator1,
        dwarf::DW_TAG_enumerator, dwarf::DW_CHILDREN_no,
        dwarf::DW_AT_name, dwarf::DW_FORM_strp,
        dwarf::DW_AT_const_value, dwarf::DW_FORM_data1,
        0, 0,

    Enumerator2,
        dwarf::DW_TAG_enumerator, dwarf::DW_CHILDREN_no,
        dwarf::DW_AT_name, dwarf::DW_FORM_strp,
        dwarf::DW_AT_const_value, dwarf::DW_FORM_data2,
        0, 0,

    Enumerator4,
        dwarf::DW_TAG_enumerator, dwarf::DW_CHILDREN_no,
        dwarf::DW_AT_name, dwarf::DW_FORM_strp,
        dwarf::DW_AT_const_value, dwarf::DW_FORM_data4,
        0, 0,

    Enumerator8,
        dwarf::DW_TAG_enumerator, dwarf::DW_CHILDREN_no,
        dwarf::DW_AT_name, dwarf::DW_FORM_strp,
        dwarf::DW_AT_const_value, dwarf::DW_FORM_data8,
        0, 0,

    TypeDef,
        dwarf::DW_TAG_typedef, dwarf::DW_CHILDREN_no,
        dwarf::DW_AT_name, dwarf::DW_FORM_strp,
        dwarf::DW_AT_type, dwarf::DW_FORM_ref4,
        0, 0,

    Subprogram,
        dwarf::DW_TAG_subprogram, dwarf::DW_CHILDREN_yes,
        dwarf::DW_AT_specification, dwarf::DW_FORM_ref4,
        dwarf::DW_AT_low_pc, dwarf::DW_FORM_addr,
        dwarf::DW_AT_high_pc, DW_FORM_size,
        dwarf::DW_AT_frame_base, dwarf::DW_FORM_exprloc,
        dwarf::DW_AT_object_pointer, dwarf::DW_FORM_ref4,
        0, 0,

    SubprogramStatic,
        dwarf::DW_TAG_subprogram, dwarf::DW_CHILDREN_yes,
        dwarf::DW_AT_specification, dwarf::DW_FORM_ref4,
        dwarf::DW_AT_low_pc, dwarf::DW_FORM_addr,
        dwarf::DW_AT_high_pc, DW_FORM_size,
        dwarf::DW_AT_frame_base, dwarf::DW_FORM_exprloc,
        0, 0,

    SubprogramSpec,
        dwarf::DW_TAG_subprogram, dwarf::DW_CHILDREN_yes,
        dwarf::DW_AT_name, dwarf::DW_FORM_strp,
        dwarf::DW_AT_linkage_name, dwarf::DW_FORM_strp,
        dwarf::DW_AT_decl_file, dwarf::DW_FORM_data1,
        dwarf::DW_AT_decl_line, dwarf::DW_FORM_data1,
        dwarf::DW_AT_type, dwarf::DW_FORM_ref4,
        dwarf::DW_AT_external, dwarf::DW_FORM_flag_present,
        dwarf::DW_AT_declaration, dwarf::DW_FORM_flag_present,
        dwarf::DW_AT_object_pointer, dwarf::DW_FORM_ref4,
        0, 0,

    SubprogramStaticSpec,
        dwarf::DW_TAG_subprogram, dwarf::DW_CHILDREN_yes,
        dwarf::DW_AT_name, dwarf::DW_FORM_strp,
        dwarf::DW_AT_linkage_name, dwarf::DW_FORM_strp,
        dwarf::DW_AT_decl_file, dwarf::DW_FORM_data1,
        dwarf::DW_AT_decl_line, dwarf::DW_FORM_data1,
        dwarf::DW_AT_type, dwarf::DW_FORM_ref4,
        dwarf::DW_AT_external, dwarf::DW_FORM_flag_present,
        dwarf::DW_AT_declaration, dwarf::DW_FORM_flag_present,
        0, 0,

    Variable,
        dwarf::DW_TAG_variable, dwarf::DW_CHILDREN_no,
        dwarf::DW_AT_name, dwarf::DW_FORM_strp,
        dwarf::DW_AT_decl_file, dwarf::DW_FORM_data1,
        dwarf::DW_AT_decl_line, dwarf::DW_FORM_data1,
        dwarf::DW_AT_type, dwarf::DW_FORM_ref4,
        dwarf::DW_AT_location, dwarf::DW_FORM_exprloc,
        0, 0,

    VariableLoc,
        dwarf::DW_TAG_variable, dwarf::DW_CHILDREN_no,
        dwarf::DW_AT_name, dwarf::DW_FORM_strp,
        dwarf::DW_AT_decl_file, dwarf::DW_FORM_data1,
        dwarf::DW_AT_decl_line, dwarf::DW_FORM_data1,
        dwarf::DW_AT_type, dwarf::DW_FORM_ref4,
        dwarf::DW_AT_location, dwarf::DW_FORM_sec_offset,
        0, 0,

    VariableStatic,
        dwarf::DW_TAG_variable, dwarf::DW_CHILDREN_no,
        dwarf::DW_AT_specification, dwarf::DW_FORM_ref4,
        dwarf::DW_AT_location, dwarf::DW_FORM_exprloc,
        0, 0,

    FormalParameter,
        dwarf::DW_TAG_formal_parameter, dwarf::DW_CHILDREN_no,
        dwarf::DW_AT_name, dwarf::DW_FORM_strp,
        dwarf::DW_AT_decl_file, dwarf::DW_FORM_data1,
        dwarf::DW_AT_decl_line, dwarf::DW_FORM_data1,
        dwarf::DW_AT_type, dwarf::DW_FORM_ref4,
        dwarf::DW_AT_location, dwarf::DW_FORM_exprloc,
        0, 0,

    FormalParameterThis,
        dwarf::DW_TAG_formal_parameter, dwarf::DW_CHILDREN_no,
        dwarf::DW_AT_name, dwarf::DW_FORM_strp,
        dwarf::DW_AT_decl_file, dwarf::DW_FORM_data1,
        dwarf::DW_AT_decl_line, dwarf::DW_FORM_data1,
        dwarf::DW_AT_type, dwarf::DW_FORM_ref4,
        dwarf::DW_AT_location, dwarf::DW_FORM_exprloc,
        dwarf::DW_AT_artificial, dwarf::DW_FORM_flag_present,
        0, 0,

    FormalParameterLoc,
        dwarf::DW_TAG_formal_parameter, dwarf::DW_CHILDREN_no,
        dwarf::DW_AT_name, dwarf::DW_FORM_strp,
        dwarf::DW_AT_decl_file, dwarf::DW_FORM_data1,
        dwarf::DW_AT_decl_line, dwarf::DW_FORM_data1,
        dwarf::DW_AT_type, dwarf::DW_FORM_ref4,
        dwarf::DW_AT_location, dwarf::DW_FORM_sec_offset,
        0, 0,

    FormalParameterThisLoc,
        dwarf::DW_TAG_formal_parameter, dwarf::DW_CHILDREN_no,
        dwarf::DW_AT_name, dwarf::DW_FORM_strp,
        dwarf::DW_AT_decl_file, dwarf::DW_FORM_data1,
        dwarf::DW_AT_decl_line, dwarf::DW_FORM_data1,
        dwarf::DW_AT_type, dwarf::DW_FORM_ref4,
        dwarf::DW_AT_location, dwarf::DW_FORM_sec_offset,
        dwarf::DW_AT_artificial, dwarf::DW_FORM_flag_present,
        0, 0,

    FormalParameterSpec,
        dwarf::DW_TAG_formal_parameter, dwarf::DW_CHILDREN_no,
        dwarf::DW_AT_type, dwarf::DW_FORM_ref4,
        0, 0,

    FormalParameterThisSpec,
        dwarf::DW_TAG_formal_parameter, dwarf::DW_CHILDREN_no,
        dwarf::DW_AT_type, dwarf::DW_FORM_ref4,
        dwarf::DW_AT_artificial, dwarf::DW_FORM_flag_present,
        0, 0,

    ClassType,
        dwarf::DW_TAG_class_type, dwarf::DW_CHILDREN_yes,
        dwarf::DW_AT_name, dwarf::DW_FORM_strp,
        dwarf::DW_AT_byte_size, dwarf::DW_FORM_data4,
        0, 0,

    ClassTypeDecl,
        dwarf::DW_TAG_class_type, dwarf::DW_CHILDREN_no,
        dwarf::DW_AT_name, dwarf::DW_FORM_strp,
        dwarf::DW_AT_declaration, dwarf::DW_FORM_flag_present,
        0, 0,

    ClassMember,
        dwarf::DW_TAG_member, dwarf::DW_CHILDREN_no,
        dwarf::DW_AT_name, dwarf::DW_FORM_strp,
        dwarf::DW_AT_type, dwarf::DW_FORM_ref4,
        dwarf::DW_AT_data_member_location, dwarf::DW_FORM_data4,
        0, 0,

    ClassMemberStatic,
        dwarf::DW_TAG_member, dwarf::DW_CHILDREN_no,
        dwarf::DW_AT_name, dwarf::DW_FORM_strp,
        dwarf::DW_AT_type, dwarf::DW_FORM_ref4,
        dwarf::DW_AT_external, dwarf::DW_FORM_flag_present,
        dwarf::DW_AT_declaration, dwarf::DW_FORM_flag_present,
        0, 0,

    PointerType,
        dwarf::DW_TAG_pointer_type, dwarf::DW_CHILDREN_no,
        dwarf::DW_AT_type, dwarf::DW_FORM_ref4,
        dwarf::DW_AT_byte_size, dwarf::DW_FORM_data1,
        0, 0,

    ReferenceType,
        dwarf::DW_TAG_reference_type, dwarf::DW_CHILDREN_no,
        dwarf::DW_AT_type, dwarf::DW_FORM_ref4,
        dwarf::DW_AT_byte_size, dwarf::DW_FORM_data1,
        0, 0,

    ArrayType,
        dwarf::DW_TAG_array_type, dwarf::DW_CHILDREN_yes,
        dwarf::DW_AT_type, dwarf::DW_FORM_ref4,
        0, 0,

    SubrangeType,
        dwarf::DW_TAG_subrange_type, dwarf::DW_CHILDREN_no,
        dwarf::DW_AT_upper_bound, dwarf::DW_FORM_udata,
        0, 0,

    ClassInheritance,
        dwarf::DW_TAG_inheritance, dwarf::DW_CHILDREN_no,
        dwarf::DW_AT_type, dwarf::DW_FORM_ref4,
        dwarf::DW_AT_data_member_location, dwarf::DW_FORM_data1,
        0, 0,

    LexicalBlock,
        dwarf::DW_TAG_lexical_block, dwarf::DW_CHILDREN_yes,
        dwarf::DW_AT_low_pc, dwarf::DW_FORM_addr,
        dwarf::DW_AT_high_pc, DW_FORM_size,
        0, 0,

    TryBlock,
        dwarf::DW_TAG_try_block, dwarf::DW_CHILDREN_no,
        dwarf::DW_AT_low_pc, dwarf::DW_FORM_addr,
        dwarf::DW_AT_high_pc, DW_FORM_size,
        0, 0,

    CatchBlock,
        dwarf::DW_TAG_catch_block, dwarf::DW_CHILDREN_no,
        dwarf::DW_AT_low_pc, dwarf::DW_FORM_addr,
        dwarf::DW_AT_high_pc, DW_FORM_size,
        0, 0,

    VoidType,
        dwarf::DW_TAG_unspecified_type, dwarf::DW_CHILDREN_no,
        dwarf::DW_AT_name, dwarf::DW_FORM_strp,
        0, 0,

    VoidPointerType,
        dwarf::DW_TAG_pointer_type, dwarf::DW_CHILDREN_no,
        0, 0,
  };

  MCContext &context = Streamer->getContext();
  Streamer->SwitchSection(context.getObjectFileInfo()->getDwarfAbbrevSection());

  for (uint16_t e : AbbrevTable) {
      Streamer->emitULEB128IntValue(e);
  }
}

}

File: src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/ObjectWriter/Dwarf/DwarfAbbrev.cs
// Licensed to the .NET Foundation under one or more agreements.
// The .NET Foundation licenses this file to you under the MIT license.

using static ILCompiler.ObjectWriter.DwarfNative;

namespace ILCompiler.ObjectWriter
{
    internal sealed class DwarfAbbrev
    {
        private readonly ushort[] _definition;

        public const ushort DW_FORM_size = 0xDEAD; // Dummy value

        private DwarfAbbrev(ushort[] definition)
        {
            _definition = definition;
        }

        public ushort Tag => _definition[0];

        public bool HasChildren => _definition[1] == DW_CHILDREN_yes;

        public void Write(SectionWriter writer, int targetPointerSize)
        {
            writer.WriteULEB128(Tag);
            writer.WriteULEB128(HasChildren ? DW_CHILDREN_yes : DW_CHILDREN_no);

            for (int i = 2; i < _definition.Length; i++)
            {
                // Attribute
                writer.WriteULEB128(_definition[i++]);
                // Form
                if (_definition[i] != DW_FORM_size)
                {
                    writer.WriteULEB128(_definition[i]);
                }
                else if (targetPointerSize == 8)
                {
                    writer.WriteULEB128(DW_FORM_data8);
                }
                else if (targetPointerSize == 4)
                {
                    writer.WriteULEB128(DW_FORM_data4);
                }
            }

            writer.Write([0, 0]);
        }

        public static readonly DwarfAbbrev CompileUnit = new([
            DW_TAG_compile_unit, DW_CHILDREN_yes,
            DW_AT_producer, DW_FORM_strp,
            DW_AT_language, DW_FORM_data2,
            DW_AT_name, DW_FORM_strp,
            DW_AT_comp_dir, DW_FORM_strp,
            DW_AT_low_pc, DW_FORM_addr,
            DW_AT_ranges, DW_FORM_sec_offset,
            DW_AT_stmt_list, DW_FORM_sec_offset]);

        public static readonly DwarfAbbrev BaseType = new([
            DW_TAG_base_type, DW_CHILDREN_no,
            DW_AT_name, DW_FORM_strp,
            DW_AT_encoding, DW_FORM_data1,
            DW_AT_byte_size, DW_FORM_data1]);

        public static readonly DwarfAbbrev EnumerationType = new([
            DW_TAG_enumeration_type, DW_CHILDREN_yes,
            DW_AT_name, DW_FORM_strp,
            DW_AT_type, DW_FORM_ref4,
            DW_AT_byte_size, DW_FORM_data1]);

        public static readonly DwarfAbbrev EnumerationTypeNoChildren = new([
            DW_TAG_enumeration_type, DW_CHILDREN_no,
            DW_AT_name, DW_FORM_strp,
            DW_AT_type, DW_FORM_ref4,
            DW_AT_byte_size, DW_FORM_data1]);

        public static readonly DwarfAbbrev Enumerator1 = new([
            DW_TAG_enumerator, DW_CHILDREN_no,
            DW_AT_name, DW_FORM_strp,
            DW_AT_const_value, DW_FORM_data1]);

        public static readonly DwarfAbbrev Enumerator2 = new([
            DW_TAG_enumerator, DW_CHILDREN_no,
            DW_AT_name, DW_FORM_strp,
            DW_AT_const_value, DW_FORM_data2]);

        public static readonly DwarfAbbrev Enumerator4 = new([
            DW_TAG_enumerator, DW_CHILDREN_no,
            DW_AT_name, DW_FORM_strp,
            DW_AT_const_value, DW_FORM_data4]);

        public static readonly DwarfAbbrev Enumerator8 = new([
            DW_TAG_enumerator, DW_CHILDREN_no,
            DW_AT_name, DW_FORM_strp,
            DW_AT_const_value, DW_FORM_data8]);

        public static readonly DwarfAbbrev TypeDef = new([
            DW_TAG_typedef, DW_CHILDREN_no,
            DW_AT_name, DW_FORM_strp,
            DW_AT_type, DW_FORM_ref4]);

        public static readonly DwarfAbbrev Subprogram = new([
            DW_TAG_subprogram, DW_CHILDREN_yes,
            DW_AT_specification, DW_FORM_ref4,
            DW_AT_low_pc, DW_FORM_addr,
            DW_AT_high_pc, DW_FORM_size,
            DW_AT_frame_base, DW_FORM_exprloc,
            DW_AT_object_pointer, DW_FORM_ref4]);

        public static readonly DwarfAbbrev SubprogramNoChildren = new([
            DW_TAG_subprogram, DW_CHILDREN_no,
            DW_AT_specification, DW_FORM_ref4,
            DW_AT_low_pc, DW_FORM_addr,
            DW_AT_high_pc, DW_FORM_size,
            DW_AT_frame_base, DW_FORM_exprloc,
            DW_AT_object_pointer, DW_FORM_ref4]);

        public static readonly DwarfAbbrev SubprogramStatic = new([
            DW_TAG_subprogram, DW_CHILDREN_yes,
            DW_AT_specification, DW_FORM_ref4,
            DW_AT_low_pc, DW_FORM_addr,
            DW_AT_high_pc, DW_FORM_size,
            DW_AT_frame_base, DW_FORM_exprloc]);

        public static readonly DwarfAbbrev SubprogramStaticNoChildren = new([
            DW_TAG_subprogram, DW_CHILDREN_no,
            DW_AT_specification, DW_FORM_ref4,
            DW_AT_low_pc, DW_FORM_addr,
            DW_AT_high_pc, DW_FORM_size,
            DW_AT_frame_base, DW_FORM_exprloc]);

        public static readonly DwarfAbbrev SubprogramSpec = new([
            DW_TAG_subprogram, DW_CHILDREN_yes,
            DW_AT_name, DW_FORM_strp,
            DW_AT_linkage_name, DW_FORM_strp,
            DW_AT_decl_file, DW_FORM_data1,
            DW_AT_decl_line, DW_FORM_data1,
            DW_AT_type, DW_FORM_ref4,
            DW_AT_external, DW_FORM_flag_present,
            DW_AT_declaration, DW_FORM_flag_present,
            DW_AT_object_pointer, DW_FORM_ref4]);

        public static readonly DwarfAbbrev SubprogramStaticSpec = new([
            DW_TAG_subprogram, DW_CHILDREN_yes,
            DW_AT_name, DW_FORM_strp,
            DW_AT_linkage_name, DW_FORM_strp,
            DW_AT_decl_file, DW_FORM_data1,
            DW_AT_decl_line, DW_FORM_data1,
            DW_AT_type, DW_FORM_ref4,
            DW_AT_external, DW_FORM_flag_present,
            DW_AT_declaration, DW_FORM_flag_present]);

        public static readonly DwarfAbbrev SubprogramStaticNoChildrenSpec = new([
            DW_TAG_subprogram, DW_CHILDREN_no,
            DW_AT_name, DW_FORM_strp,
            DW_AT_linkage_name, DW_FORM_strp,
            DW_AT_decl_file, DW_FORM_data1,
            DW_AT_decl_line, DW_FORM_data1,
            DW_AT_type, DW_FORM_ref4,
            DW_AT_external, DW_FORM_flag_present,
            DW_AT_declaration, DW_FORM_flag_present]);

        public static readonly DwarfAbbrev Variable = new([
            DW_TAG_variable, DW_CHILDREN_no,
            DW_AT_name, DW_FORM_strp,
            DW_AT_decl_file, DW_FORM_data1,
            DW_AT_decl_line, DW_FORM_data1,
            DW_AT_type, DW_FORM_ref4,
            DW_AT_location, DW_FORM_exprloc]);

        public static readonly DwarfAbbrev VariableLoc = new([
            DW_TAG_variable, DW_CHILDREN_no,
            DW_AT_name, DW_FORM_strp,
            DW_AT_decl_file, DW_FORM_data1,
            DW_AT_decl_line, DW_FORM_data1,
            DW_AT_type, DW_FORM_ref4,
            DW_AT_location, DW_FORM_sec_offset]);

        public static readonly DwarfAbbrev VariableStatic = new([
            DW_TAG_variable, DW_CHILDREN_no,
            DW_AT_specification, DW_FORM_ref4,
            DW_AT_location, DW_FORM_exprloc]);

        public static readonly DwarfAbbrev FormalParameter = new([
            DW_TAG_formal_parameter, DW_CHILDREN_no,
            DW_AT_name, DW_FORM_strp,
            DW_AT_decl_file, DW_FORM_data1,
            DW_AT_decl_line, DW_FORM_data1,
            DW_AT_type, DW_FORM_ref4,
            DW_AT_location, DW_FORM_exprloc]);

        public static readonly DwarfAbbrev FormalParameterThis = new([
            DW_TAG_formal_parameter, DW_CHILDREN_no,
            DW_AT_name, DW_FORM_strp,
            DW_AT_decl_file, DW_FORM_data1,
            DW_AT_decl_line, DW_FORM_data1,
            DW_AT_type, DW_FORM_ref4,
            DW_AT_location, DW_FORM_exprloc,
            DW_AT_artificial, DW_FORM_flag_present]);

        public static readonly DwarfAbbrev FormalParameterLoc = new([
            DW_TAG_formal_parameter, DW_CHILDREN_no,
            DW_AT_name, DW_FORM_strp,
            DW_AT_decl_file, DW_FORM_data1,
            DW_AT_decl_line, DW_FORM_data1,
            DW_AT_type, DW_FORM_ref4,
            DW_AT_location, DW_FORM_sec_offset]);

        public static readonly DwarfAbbrev FormalParameterThisLoc = new([
            DW_TAG_formal_parameter, DW_CHILDREN_no,
            DW_AT_name, DW_FORM_strp,
            DW_AT_decl_file, DW_FORM_data1,
            DW_AT_decl_line, DW_FORM_data1,
            DW_AT_type, DW_FORM_ref4,
            DW_AT_location, DW_FORM_sec_offset,
            DW_AT_artificial, DW_FORM_flag_present]);

        public static readonly DwarfAbbrev FormalParameterSpec = new([
            DW_TAG_formal_parameter, DW_CHILDREN_no,
            DW_AT_type, DW_FORM_ref4]);

        public static readonly DwarfAbbrev FormalParameterThisSpec = new([
            DW_TAG_formal_parameter, DW_CHILDREN_no,
            DW_AT_type, DW_FORM_ref4,
            DW_AT_artificial, DW_FORM_flag_present]);

        public static readonly DwarfAbbrev ClassType = new([
            DW_TAG_class_type, DW_CHILDREN_yes,
            DW_AT_name, DW_FORM_strp,
            DW_AT_byte_size, DW_FORM_data4]);

        public static readonly DwarfAbbrev ClassTypeNoChildren = new([
            DW_TAG_class_type, DW_CHILDREN_no,
            DW_AT_name, DW_FORM_strp,
            DW_AT_byte_size, DW_FORM_data4]);

        public static readonly DwarfAbbrev ClassTypeDecl = new([
            DW_TAG_class_type, DW_CHILDREN_no,
            DW_AT_name, DW_FORM_strp,
            DW_AT_declaration, DW_FORM_flag_present]);

        public static readonly DwarfAbbrev ClassMember = new([
            DW_TAG_member, DW_CHILDREN_no,
            DW_AT_name, DW_FORM_strp,
            DW_AT_type, DW_FORM_ref4,
            DW_AT_data_member_location, DW_FORM_data4]);

        public static readonly DwarfAbbrev ClassMemberStatic = new([
            DW_TAG_member, DW_CHILDREN_no,
            DW_AT_name, DW_FORM_strp,
            DW_AT_type, DW_FORM_ref4,
            DW_AT_external, DW_FORM_flag_present,
            DW_AT_declaration, DW_FORM_flag_present]);

        public static readonly DwarfAbbrev PointerType = new([
            DW_TAG_pointer_type, DW_CHILDREN_no,
            DW_AT_type, DW_FORM_ref4,
            DW_AT_byte_size, DW_FORM_data1]);

        public static readonly DwarfAbbrev ReferenceType = new([
            DW_TAG_reference_type, DW_CHILDREN_no,
            DW_AT_type, DW_FORM_ref4,
            DW_AT_byte_size, DW_FORM_data1]);

        public static readonly DwarfAbbrev ArrayType = new([
            DW_TAG_array_type, DW_CHILDREN_yes,
            DW_AT_type, DW_FORM_ref4]);

        public static readonly DwarfAbbrev SubrangeType = new([
            DW_TAG_subrange_type, DW_CHILDREN_no,
            DW_AT_upper_bound, DW_FORM_udata]);

        public static readonly DwarfAbbrev ClassInheritance = new([
            DW_TAG_inheritance, DW_CHILDREN_no,
            DW_AT_type, DW_FORM_ref4,
            DW_AT_data_member_location, DW_FORM_data1]);

        public static readonly DwarfAbbrev LexicalBlock = new([
            DW_TAG_lexical_block, DW_CHILDREN_yes,
            DW_AT_low_pc, DW_FORM_addr,
            DW_AT_high_pc, DW_FORM_size]);

        public static readonly DwarfAbbrev TryBlock = new([
            DW_TAG_try_block, DW_CHILDREN_no,
            DW_AT_low_pc, DW_FORM_addr,
            DW_AT_high_pc, DW_FORM_size]);

        public static readonly DwarfAbbrev CatchBlock = new([
            DW_TAG_catch_block, DW_CHILDREN_no,
            DW_AT_low_pc, DW_FORM_addr,
            DW_AT_high_pc, DW_FORM_size]);

        public static readonly DwarfAbbrev VoidType = new([
            DW_TAG_unspecified_type, DW_CHILDREN_no,
            DW_AT_name, DW_FORM_strp]);

        public static readonly DwarfAbbrev VoidPointerType = new([
            DW_TAG_pointer_type, DW_CHILDREN_no]);
    }
}
